src/swap/switchain.js

```diff
@@ -0,0 +1,371 @@
+}
+
+let apiUrl = 'https://api.switchain.com/rest/v1'
+
+type SwitchainResponseError = {
+  error: string,
+  reason?: string
+}
+
+type SwitchainOfferResponse = {
+  pair: string,
+  signature: string,
+  quote: string,
+  maxLimit: string,
+  minLimit: string,
+  expiryTs: number,
+  minerFee: string,
+  orderId?: string
+}
+
+type SwitchainOrderCreationResponse = {
+  orderId: string,
+  fromAmount: string,
+  rate: string,
+  exchangeAddress: string,
+  exchangeAddressTag?: string,
+  refundAddress: string,
+  refundAddressTag?: string,
+  toAddres: string,
+  toAddressTag?: string
+}
+
+type SwitchainOrderCreationBody = {
+  pair: string,
+  toAddress: string,
+  toAddressTag?: string,
+  refundAddress: string,
+  refundAddressTag?: string,
+  signature: string,
+  fromAmount?: string,
+  toAmount?: string
+}
+
+const dontUseLegacy = {
+  DGB: true
+}
+
+export function makeSwitchainPlugin(
+  opts: EdgeCorePluginOptions
+): EdgeSwapPlugin {
+  const { initOptions, io } = opts
+
+  if (!initOptions.apiKey) {
+    throw new Error('No Switchain API key provided.')
+  }
+
+  async function getAddress(
+    wallet: EdgeCurrencyWallet,
+    currencyCode: string
+  ): Promise<string> {
+    const addressInfo = await wallet.getReceiveAddress({ currencyCode })
+
+    return addressInfo.legacyAddress && !dontUseLegacy[currencyCode]
+      ? addressInfo.legacyAddress
+      : addressInfo.publicAddress
+  }
+
+  async function swHttpCall(
+    path: string,
+    method: string,
+    body?: Object,
+    query?: { [string]: string }
+  ) {
+    let requestOpts = {
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: `Bearer ${initOptions.apiKey}`
+      },
+      method
+    }
+
+    if (body) {
+      requestOpts = { ...requestOpts, body: JSON.stringify(body) }
+    }
+    let queryParams = ''
+    if (query) {
+      // only way i found to circumvent Object.entries type errors
+      const queryStringList: Array<[string, string]> = Object.entries(
+        query
+      ).map(([k, v]) => [k, String(v)])
+      queryParams = `?${new URLSearchParams(queryStringList).toString()}`
+    }
+    const uri = `${apiUrl}${path}${queryParams}`
+    const reply: EdgeFetchResponse = await io.fetch(uri, requestOpts)
+
+    let replyJson:
+      | SwitchainOfferResponse
+      | SwitchainOrderCreationResponse
+      | SwitchainResponseError
+    try {
+      replyJson = await reply.json()
+    } catch (e) {
+      throw new Error(
+        `Switchain ${uri} returned error code ${reply.status} (no JSON)`
+      )
+    }
+
+    if (reply.status !== 200) {
+      if (replyJson.reason) {
+        throw new Error(replyJson.reason)
+      }
+
+      throw new Error(
+        `Switchain ${uri} returned error code ${
+          replyJson.status
+        } with JSON ${JSON.stringify(replyJson)}`
+      )
+    }
+
+    return replyJson
+  }
+
+  // closures to be populated between `fetchSwapQuote` and `approve`
+  let fromCurrencyCode = ''
```

Comment:
If you inline `makeSwapPluginQuote` into `fetchSwapQuote`, you won't need these closure variables. Might be cleaner, but I won't insist on a big refactor if you don't want to.

Reply:
Only the `orderCreationBody` closure remains as I need to bridge `fetchSwapQuote` and `makeSwapPluginQuote`.